File: mobile_flutter/lib/services/location_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../app_config.dart';

/// Reads the country / state / city lists proxied by the Express backend.
class LocationService {
  final String _base = AppConfig.baseUrl;

  Future<List<Map<String, dynamic>>> getCountries(String token) async {
    final res = await http.get(
      Uri.parse('$_base/locations/countries'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (res.statusCode == 200) {
      final body = jsonDecode(res.body);
      return List<Map<String, dynamic>>.from(body['data']);
    }
    throw Exception('Failed to load countries');
  }

  Future<List<Map<String, dynamic>>> getStates(String token, String ciso) async {
    final res = await http.get(
      Uri.parse('$_base/locations/countries/$ciso/states'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (res.statusCode == 200) {
      final body = jsonDecode(res.body);
      return List<Map<String, dynamic>>.from(body['data']);
    }
    throw Exception('Failed to load states');
  }

  Future<List<Map<String, dynamic>>> getCities(
      String token, String ciso, String siso) async {
    final res = await http.get(
      Uri.parse('$_base/locations/countries/$ciso/states/$siso/cities'),
      headers: {'Authorization': 'Bearer $token'},
    );
    if (res.statusCode == 200) {
      final body = jsonDecode(res.body);
      return List<Map<String, dynamic>>.from(body['data']);
    }
    throw Exception('Failed to load cities');
  }
}

File: backend_express/src/services/locationService.js
const axios = require('axios');

const CSC_BASE_URL = 'https://api.countrystatecity.in/v1';

const request = async (path) => {
  const response = await axios.get(`${CSC_BASE_URL}${path}`, {
    headers: {
      'X-CSCAPI-KEY': process.env.CSC_API_KEY,
    },
  });
  return response.data;
};

const getCountries = async () => request('/countries');

const getStates = async (ciso) =>
  request(`/countries/${encodeURIComponent(ciso)}/states`);

const getCities = async (ciso, siso) =>
  request(
    `/countries/${encodeURIComponent(ciso)}/states/${encodeURIComponent(siso)}/cities`
  );

module.exports = { getCountries, getStates, getCities };
